openforcefield/typing/engines/smirnoff/parameters.py

```diff
@@ -2219,12 +2224,78 @@ def create_force(self, system, topology, **kwargs):
                 len(improper_matches)))
 
 
-class vdWHandler(ParameterHandler):
+class _NonbondedHandler(ParameterHandler):
+    """Base class for ParameterHandlers that deal with OpenMM NonbondedForce objects."""
+    _OPENMMTYPE = openmm.NonbondedForce
+
+    def create_force(self, system, topology, **kwargs):
+        # If we aren't yet keeping track of which molecules' charges have been assigned by which charge methods,
+        # initialize a dict for that here.
+        # TODO: This should be an attribute of the _system_, not the _topology_. However, since we're still using
+        #  OpenMM's System class, I am storing this data on the OFF Topology until we make an OFF System class.
+        if not hasattr(topology, "_ref_mol_to_charge_method"):
+            topology._ref_mol_to_charge_method = {ref_mol: None for ref_mol in topology.reference_molecules}
```

Comment:
I wonder if rather than tagging the topology directly with this attribute, it would be cleaner to carry around some additional, extensible, internal state object which gets passed between successive handlers and that can store intermediate data such as this?

Reply:
This is admittedly a cheap hack. The object which this information *should* be stored on is the `System` (because then after parameterization, users will be able to look right at a particle and see what ParameterHandler the charge came from). 

I don't know how I feel about introducing an internal state object, so I'm thinking (typing) out loud here... It seems like this would partially be a "parametrization event" object, effectively like a rich log of what happened. This might be good for tracking provenance -- and actually we just talked about this at the MolSSI workshop (It would be neat if system prep tools produced an output which capture the important details of what they've done, which could be submitted as SI, relevant to @karmencj ). So I can envision it as purely an object to be written _to_. However, becomes more complex if it passes information _back_ to the FF, topology, or system during the parametrization process (which seems to be the proposal here). Unless there's some clear model of when it talks to the other objects, we could end up with some very hard-to-test behaviors. I'll keep thinking about how this could work, though.

Reply:
> object which this information should be stored on is the System
This may be the case - although I'm not sure what the status of this object is?

> This might be good for tracking provenance
In principle it would - it's never a bad thing to keep track of what went into and came out of the different modular handlers.

> However, becomes more complex if it passes information back to the FF, topology, or system during the parametrization process [...] Unless there's some clear model of when it talks to the other objects

I don't think I understand this point - where would this complexity arise? The object wouldn't need to talk to anything, it would purely be an object which gets passed between handlers which would be a readable source of previously assigned metadata which could either be utilized if needed, or ignored if not, and then extended ready to be passed along the chain of handlers. In principle the object could be as simple as a dict of dicts, where each key is the handler name and the value the dict of provenance.

> we could end up with some very hard-to-test behaviors

I think that is inherent in having modular components which can be almost arbitrarily pieced together, but provided the required input and provided output provenance of each handler is well defined, then it shouldn't be overly bad!

Honestly I don't have particularly strong opinions about this, but given that in the future there is probably going to be cases where handlers may need access to information about what has already been done to the system (as is the case here), it might be worth thinking about how this will worker earlier rather than later!

Reply:
I'd be very careful about decorating the source `Topology` object with something that should change when the same `Topology` object is fed to a different `ForceField`. 

I do think we should seriously consider a new openforcefield `System` object as the right way to contain all of the applied parameterization data accessible via a public API.